manticore/core/smtlib/solver.py

```diff
@@ -143,207 +167,162 @@ def minmax(self, constraints, x, iters=10000):
-            print(e, "Probably too many cached expressions? visitors._cache...")
-            # Z3 was removed from the system in the middle of operation
-            raise Z3NotFoundError  # TODO(mark) don't catch this exception in two places
-
-        # run solver specific initializations
-        for cfg in self._init:
-            self._send(cfg)
+    def start(self):
+        """Spawns POpen solver process"""
+        if self._proc is not None:
+            return
+        self._proc = Popen(
+            shlex.split(self._command),
+            stdin=PIPE,
+            stdout=PIPE,
+            bufsize=0,
+            universal_newlines=True,
+            close_fds=True,
+        )
 
-    def _stop_proc(self):
+    def stop(self):
         """
-        Stops the z3 solver process by:
-        - sending an exit command to it,
+        Stops the solver process by:
         - sending a SIGKILL signal,
         - waiting till the process terminates (so we don't leave a zombie process)
         """
         if self._proc is None:
             return
+        # if it did not finished already
         if self._proc.returncode is None:
-            try:
-                self._send("(exit)")
-            except (SolverError, IOError) as e:
-                # z3 was too fast to close
-                logger.debug(str(e))
-            finally:
-                try:
-                    self._proc.stdin.close()
-                except IOError as e:
-                    logger.debug(str(e))
-                try:
-                    self._proc.stdout.close()
-                except IOError as e:
-                    logger.debug(str(e))
-                self._proc.kill()
-                # Wait for termination, to avoid zombies.
-                self._proc.wait()
-
-        self._proc: Popen = None
-
-    # marshaling/pickle
-    def __getstate__(self):
-        raise SolverException()
-
-    def __setstate__(self, state):
-        raise SolverException()
-
-    def __del__(self):
-        try:
-            if self._proc is not None:
-                self._stop_proc()
-            # self._proc.stdin.writelines(('(exit)\n',))
-            # self._proc.wait()
-        except Exception as e:
-            logger.error(str(e))
-
-    def _reset(self, constraints: Optional[str] = None) -> None:
-        """Auxiliary method to reset the smtlib external solver to initial defaults"""
-        if self._proc is None:
-            self._start_proc()
-        else:
-            if self.support_reset:
-                self._send("(reset)")
-
-                for cfg in self._init:
-                    self._send(cfg)
-            else:
-                self._stop_proc()
-                self._start_proc()
-        if constraints is not None:
-            self._send(constraints)
-
-    def _send(self, cmd: str) -> None:
+            self._proc.stdin.close()
+            self._proc.stdout.close()
+            # Kill the process
+            self._proc.kill()
```

Comment:
As long as we're going to wait for termination, shouldn't we send SIGTERM first instead of SIGKILL? From the comment, it looks like that might have been removed. Any reason why?

Reply:
You mean that the _proc.wait() is unnecessary after a _proc.kill()?
I do not trust that the .kill() translates to a specific signal.
I want to destroy the solver companion the easiest, fastest way.

Reply:
From the docs, on Linux/Mac, [`.terminate`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.terminate) sends SIGTERM, which gives the process a chance to clean up first, and [`.kill`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.kill) sends a SIGKILL, which yanks the rug out from under the process and kills it instantly. So, if we're using `.kill`, there's no need for `.wait`. That said, it would be slightly safer to do something like:
```
self._proc.terminate()
try:
    self._proc.wait(2)
except TimeoutExpired:
    self._proc.kill()
```

Reply:
Removed the .wait. As supposedly kill() does it all. If we start seeing zombies we'll add some code here.